ckanext/spectrum/actions.py

```diff
@@ -61,6 +64,70 @@ def user_create(next_action, context, data_dict):
     return created_user
 
 
+def dataset_tag_patch(context, data_dict):
+    if 'tags' not in data_dict or not isinstance(data_dict['tags'], dict):
+        raise toolkit.ValidationError(toolkit._(
+            "Must specify 'tags' dict of tags for update in form "
+            "{'old_tag_name1': 'new_tag_name1', 'old_tag_name2': 'new_tag_name2'}"))
+
+    package_search_params = {}
+    for key in allowed_package_search_params:
+        if key in data_dict:
+            package_search_params[key] = data_dict[key]
+
+    package_search_params = _restrict_datasets_to_those_with_tags(package_search_params, data_dict['tags'])
+
+    datasets = toolkit.get_action('package_search')(context, package_search_params)
+
+    if _user_has_access_to_all_datasets(context, datasets):
+        _update_tags(context, datasets, data_dict['tags'])
+    else:
+        raise toolkit.NotAuthorized(toolkit._("User has no access for patching some datasets"))
+
+    return {'datasets_modified': datasets['count']}
+
+
+def _user_has_access_to_all_datasets(context, datasets):
+    for ds in datasets["results"]:
+        if not toolkit.check_access('package_patch', context, {"id": ds['id']}):
+            return False
+
+    return True
+
+
+def _restrict_datasets_to_those_with_tags(package_search_params, tags):
+    fq_tag_restriction = _create_fq_tag_restriction(tags)
+
+    if 'fq' in package_search_params:
+        original_fq = package_search_params['fq']
+        package_search_params['fq'] = f"({original_fq}) AND ({fq_tag_restriction})"
+    else:
+        package_search_params['fq'] = f"({fq_tag_restriction})"
+
+    return package_search_params
+
+
+def _update_tags(context, datasets, tags_to_be_replaced):
+    dataset_patch_action = toolkit.get_action('package_patch')
+
+    for ds in datasets["results"]:
+        final_tags = _prepare_final_tag_list(ds['tags'], tags_to_be_replaced)
+        dataset_patch_action(context, {'id': ds['id'], 'tags': final_tags})
+
+
+def _prepare_final_tag_list(original_tags, tags_to_be_replaced):
+    final_tags = []
+    for tag in original_tags:
+        tag_name = tag['name']
+        final_tags.append({"name": tags_to_be_replaced[tag_name] if tag_name in tags_to_be_replaced else tag_name})
+
+    return final_tags
+
+
+def _create_fq_tag_restriction(tags):
+    return " OR ".join([f"tags:{key}" for key in tags])
```

Comment:
Did we agree to just restrict one request to updating one tag?

To be honest I think I would still prefer the one-request-one-tag approach, as I see it, it would create much simpler code to maintain, simpler api interface and a simpler mental model for how to use the endpoint. I don't think we make big efficiency savings by allowing the user to update multiple tags at once and I also don't know that the client will ever need to update multiple tags at once. 

However, if what you have works now, and is tested, since we are on a tight schedule, I don't have a problem if you would rather leave as it is. 

If you were only updating a single tag, you could probably reduce the code above like so...

```suggestion
def _update_tags(context, datasets, old_tag, new_tag):
    for dataset in datasets["results"]:
        updated_tags = [
            tag for tag in dataset["tags"] if tag["name"] != old_tag
         ] + [{"name": new_tag}]
         toolkit.get_action('package_patch')(
             context, 
             {'id': dataset['id'], 'tags': updated_tags}
         )
```

Reply:
Yep, we did agree, but I found out that whether you pass tags in request parameters as [{oldtag1: newtag1},{oldtag2: newtag2}] or {oldtag1: newtag1, oldtag2: newtag2} or {oldtag1: newtag1} in the end you get a dict with all those keys and values. If we were to allow only 1 tag to change, we'd need here to check if there is only 1 key passed. This check would add complexity (however in _update_tags we have slightly more complex code on the other hand) so I decided to leave it. I'd suggest leaving as is.